flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/window/utils/WindowJoinHelper.java

```diff
@@ -0,0 +1,402 @@
+import org.apache.flink.table.runtime.operators.window.tvf.common.WindowTimerService;
+import org.apache.flink.table.runtime.typeutils.RowDataSerializer;
+import org.apache.flink.types.RowKind;
+
+import javax.annotation.Nullable;
+
+import java.time.ZoneId;
+import java.util.IdentityHashMap;
+
+import static org.apache.flink.table.runtime.util.TimeWindowUtil.isWindowFired;
+
+/**
+ * A helper to do the window join operations for {@link WindowJoinOperator} and {@link
+ * AsyncStateWindowJoinOperator}.
+ */
+public abstract class WindowJoinHelper {
+
+    private static final String LEFT_LATE_ELEMENTS_DROPPED_METRIC_NAME =
+            "leftNumLateRecordsDropped";
+    private static final String LEFT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME =
+            "leftLateRecordsDroppedRate";
+    private static final String RIGHT_LATE_ELEMENTS_DROPPED_METRIC_NAME =
+            "rightNumLateRecordsDropped";
+    private static final String RIGHT_LATE_ELEMENTS_DROPPED_RATE_METRIC_NAME =
+            "rightLateRecordsDroppedRate";
+    private static final String WATERMARK_LATENCY_METRIC_NAME = "watermarkLatency";
+
+    private final ZoneId shiftTimeZone;
+
+    private final WindowTimerService<Long> windowTimerService;
+
+    protected final RowDataSerializer leftSerializer;
+
+    protected final RowDataSerializer rightSerializer;
+
+    protected final JoinConditionWithNullFilters joinCondition;
+
+    private final int leftWindowEndIndex;
+    private final int rightWindowEndIndex;
+
+    /** This is used for emitting elements with a given timestamp. */
+    protected final TimestampedCollector<RowData> collector;
+
+    private final WindowJoinProcessor windowJoinProcessor;
+
+    // ------------------------------------------------------------------------
+    // Metrics
+    // ------------------------------------------------------------------------
+
+    private final Meter leftLateRecordsDroppedRate;
+    private final Meter rightLateRecordsDroppedRate;
+
+    public WindowJoinHelper(
+            RowDataSerializer leftSerializer,
+            RowDataSerializer rightSerializer,
+            ZoneId shiftTimeZone,
+            WindowTimerService<Long> windowTimerService,
+            JoinConditionWithNullFilters joinCondition,
+            int leftWindowEndIndex,
+            int rightWindowEndIndex,
+            TimestampedCollector<RowData> collector,
+            FlinkJoinType joinType,
+            OperatorMetricGroup metrics) {
+        this.leftSerializer = leftSerializer;
+        this.rightSerializer = rightSerializer;
+        this.shiftTimeZone = shiftTimeZone;
+        this.windowTimerService = windowTimerService;
+        this.joinCondition = joinCondition;
+        this.leftWindowEndIndex = leftWindowEndIndex;
+        this.rightWindowEndIndex = rightWindowEndIndex;
+        this.collector = collector;
+
+        this.windowJoinProcessor = getWindowJoinProcessor(joinType);
+
+        // register metrics
```

Comment:
We commonly place the metric registrations in the open/setup methods, but it is a bit obscure to place it in the constructor, WDYT?

Reply:
Done